src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,198 @@
+    }
+
+    /**
+    * @notice Called by a staker or operator to prove the inclusion of their earnings in a posted Merkle root and claim them.
+    * @param proof Merkle proof showing that a leaf was included in the `rootIndex`-th
+    * Merkle root posted for the `token`
+    * @param rootIndex Specifies the node inside the Merkle tree corresponding to the specified root, `merkleRoots[rootIndex].root`.
+    */
+    function proveAndClaimEarnings(
+        bytes memory proof,
+        uint256 rootIndex,
+        MerkleLeaf memory leaf,
+        uint256 leafIndex
+    ) external {
+        require(leaf.amounts.length == leaf.tokens.length, "PaymentCoordinator.proveAndClaimEarnings: leaf amounts and tokens must be same length");
+        require(_merkleRootPosts[rootIndex].confirmedAtBlockNumber < block.number, "PaymentCoordinator.proveAndClaimEarnings: Merkle root not yet confirmed");
+
+        bytes32 root = _merkleRootPosts[rootIndex].root;
+        require(root != bytes32(0), "PaymentCoordinator.proveAndClaimEarnings: Merkle root is null");
+
+        bytes32 leafHash = computeLeafHash(leaf);
+        require(Merkle.verifyInclusionKeccak(proof, root, leafHash, leafIndex), "PaymentCoordinator.proveAndClaimEarnings: Invalid proof");
+
+        for(uint256 i = 0; i < leaf.amounts.length; i++) {
+            uint256 amount = leaf.amounts[i] - cumulativeTokenAmountClaimedByRecipient[leaf.tokens[i]][leaf.recipient];
+            cumulativeTokenAmountClaimedByRecipient[leaf.tokens[i]][leaf.recipient] = leaf.amounts[i];
+            leaf.tokens[i].safeTransfer(leaf.recipient, amount);
+            
+        }
+        emit PaymentClaimed(leaf);
+    }
+
+    function setRootPublisher(address _rootPublisher) external onlyOwner {
+        _setRootPublisher(_rootPublisher);
+    }
+
+    function setMerkleRootActivationDelayBlocks(uint256 _merkleRootActivationDelayBlocks) external onlyOwner {
+        _setMerkleRootActivationDelayBlocks(_merkleRootActivationDelayBlocks);
+    }
+
+    function setEigenLayerShareBIPS(uint256 _eigenlayerShareBips) external onlyOwner {
+        _setEigenLayerShareBIPS(_eigenlayerShareBips);
+    }
+
+
+    /// @notice Getter function for the length of the `merkleRoots` array
+    function merkleRootPostsLength() external view returns (uint256) {
+        return _merkleRootPosts.length;
+    }
+
+    /// @notice Getter function for a merkleRoot at a given index
+    function merkleRootPosts(uint256 index) external view returns (MerkleRootPost memory) {
+        return _merkleRootPosts[index];
+    }
+
+    function _setRootPublisher(address _rootPublisher) internal {
+        address currentRootPublisher = rootPublisher;
+        rootPublisher = _rootPublisher;
+        emit RootPublisherChanged(currentRootPublisher, rootPublisher);
+    }
+
+    function _setMerkleRootActivationDelayBlocks(uint256 _merkleRootActivationDelayBlocks) internal {
+        merkleRootActivationDelayBlocks = _merkleRootActivationDelayBlocks;
```

Comment:
probably want some bounds on this check.
I am OK leaving it for the moment but I think we should have a `TODO` comment at least